mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java

```diff
@@ -197,42 +198,64 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
     }
 
     @Override
-    public Map<MessageUid, MessageMetaData> expungeMarkedForDeletionInMailbox(Mailbox mailbox, MessageRange set)
-            throws MailboxException {
+    public List<MessageUid> retrieveMessagesMarkedForDeletion(Mailbox mailbox, MessageRange messageRange) throws MailboxException {
         try {
-            final Map<MessageUid, MessageMetaData> data;
-            final List<MailboxMessage> results;
-            final MessageUid from = set.getUidFrom();
-            final MessageUid to = set.getUidTo();
+            MessageUid from = messageRange.getUidFrom();
+            MessageUid to = messageRange.getUidTo();
             JPAId mailboxId = (JPAId) mailbox.getMailboxId();
+            List<MailboxMessage> messages = findDeletedMessages(messageRange, from, to, mailboxId);
+            return getUidList(messages);
+        } catch (PersistenceException e) {
+            throw new MailboxException("Search of MessageRange " + messageRange + " failed in mailbox " + mailbox, e);
+        }
+    }
+
+    @Override
+    public Map<MessageUid, MessageMetaData> deleteMessages(Mailbox mailbox, List<MessageUid> uids) throws MailboxException {
+        JPAId mailboxId = (JPAId) mailbox.getMailboxId();
+        Map<MessageUid, MessageMetaData> data = new HashMap<>();
+        List<MessageRange> ranges = MessageRange.toRanges(uids);
+
+        ranges.forEach(range -> {
+            MessageUid from = range.getUidFrom();
+            MessageUid to = range.getUidTo();
+            List<MailboxMessage> messages = findDeletedMessages(range, from, to, mailboxId);
+            data.putAll(createMetaData(messages));
+            deleteDeletedMessages(range, from, to, mailboxId);
+        });
+
+        return data;
+    }
+
+    private List<MailboxMessage> findDeletedMessages(MessageRange messageRange, MessageUid from, MessageUid to, JPAId mailboxId) {
```

Comment:
I think this method only need one parameter `MessageRange`, the others `from` and `to` can be easily retrieved from the range.